crates/re_query_cache/src/query.rs

```diff
@@ -0,0 +1,260 @@
+use paste::paste;
+use seq_macro::seq;
+
+use re_data_store::{DataStore, LatestAtQuery, TimeInt, Timeline};
+use re_entity_db::{ExtraQueryHistory, VisibleHistory};
+use re_log_types::{EntityPath, RowId};
+use re_query::query_archetype;
+use re_types_core::{components::InstanceKey, Archetype, Component};
+
+use crate::{AnyQuery, Caches};
+
+// ---
+
+/// Either a reference to a slice of data residing in the cache, or some data being deserialized
+/// just-in-time from an [`re_query::ArchetypeView`].
+#[derive(Debug, Clone)]
+pub enum MaybeCachedComponentData<'a, C> {
+    Cached(&'a [C]),
+    // TODO(cmc): Ideally, this would be a reference to a `dyn Iterator` that is the result of
+    // calling `ArchetypeView::iter_{required|optional}_component`.
+    // In practice this enters lifetime invariance hell for, from what I can see, no particular gains
+    // (rustc is pretty good at optimizing out collections into obvious temporary variables).
+    Raw(Vec<C>),
+}
+
+impl<'a, C: Clone> MaybeCachedComponentData<'a, C> {
+    #[inline]
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = &C> + '_ {
+        self.as_slice().iter()
+    }
+
+    #[inline]
+    pub fn as_slice(&self) -> &[C] {
```

Comment:
We should also implement `Deref for MaybeCachedComponentData` with `type Target = [C];`